polkadot/primitives/src/vstaging/mod.rs

```diff
@@ -310,41 +314,54 @@ pub enum UMPSignal {
 pub const UMP_SEPARATOR: Vec<u8> = vec![];
 
 impl CandidateCommitments {
-	/// Returns the core selector and claim queue offset the candidate has committed to, if any.
-	pub fn selected_core(&self) -> Option<(CoreSelector, ClaimQueueOffset)> {
-		// We need at least 2 messages for the separator and core selector
-		if self.upward_messages.len() < 2 {
-			return None
-		}
-
-		let separator_pos =
-			self.upward_messages.iter().rposition(|message| message == &UMP_SEPARATOR)?;
-
-		// Use first commitment
-		let message = self.upward_messages.get(separator_pos + 1)?;
-
-		match UMPSignal::decode(&mut message.as_slice()).ok()? {
-			UMPSignal::SelectCore(core_selector, cq_offset) => Some((core_selector, cq_offset)),
-		}
-	}
-
 	/// Returns the core index determined by `UMPSignal::SelectCore` commitment
 	/// and `assigned_cores`.
 	///
 	/// Returns `None` if there is no `UMPSignal::SelectCore` commitment or
 	/// assigned cores is empty.
 	///
 	/// `assigned_cores` must be a sorted vec of all core indices assigned to a parachain.
-	pub fn committed_core_index(&self, assigned_cores: &[&CoreIndex]) -> Option<CoreIndex> {
+	pub fn committed_core_index(
+		&self,
+		assigned_cores: &BTreeMap<u8, BTreeSet<CoreIndex>>,
+	) -> Result<CoreIndex, CandidateReceiptError> {
+		let mut signals_iter =
+			self.upward_messages.iter().skip_while(|message| *message != &UMP_SEPARATOR);
+
+		let (core_selector, cq_offset) = if signals_iter.next().is_some() {
+			let core_selector_message =
+				signals_iter.next().ok_or(CandidateReceiptError::NoCoreSelected)?;
+			// We should have exactly one signal beyond the separator
+			if signals_iter.next().is_some() {
+				return Err(CandidateReceiptError::TooManyUMPSignals)
+			}
+
+			match UMPSignal::decode(&mut core_selector_message.as_slice())
+				.map_err(|_| CandidateReceiptError::InvalidSelectedCore)?
+			{
+				UMPSignal::SelectCore(core_selector, cq_offset) => (Some(core_selector), cq_offset),
+			}
+		} else {
+			// No separator, we should use the defaults.
+			(None, ClaimQueueOffset(DEFAULT_CLAIM_QUEUE_OFFSET))
+		};
+
+		let assigned_cores =
+			assigned_cores.get(&cq_offset.0).ok_or(CandidateReceiptError::NoAssignment)?;
+
 		if assigned_cores.is_empty() {
-			return None
+			return Err(CandidateReceiptError::NoAssignment)
 		}
 
-		self.selected_core().and_then(|(core_selector, _cq_offset)| {
-			let core_index =
-				**assigned_cores.get(core_selector.0 as usize % assigned_cores.len())?;
-			Some(core_index)
-		})
+		let core_selector = core_selector
+			.or_else(|| if assigned_cores.len() == 1 { Some(CoreSelector(0)) } else { None })
```

Comment:
We need to still allow v2 candidates without core selectors, only if the MVP bit is enabled, because they will have an injected core index and will be valid candidates in the runtime.

Reply:
done